### DISCUSSION TITLE
Type Hint the Python 3.8 LRU Cache Example

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [python/typing#1253] [Q&A] [answered]
Can someone please provide me guidance as to the best (recommended) way to type hint the `LRU` cache example from the [Python 3.8 docs](https://docs.python.org/3.8/library/collections.html#ordereddict-examples-and-recipes). To reiterate, the example source is:

```python
class LRU(OrderedDict):
    'Limit size, evicting the least recently looked-up key when full'

    def __init__(self, maxsize=128, /, *args, **kwds):
        self.maxsize = maxsize
        super().__init__(*args, **kwds)

    def __getitem__(self, key):
        value = super().__getitem__(key)
        self.move_to_end(key)
        return value

    def __setitem__(self, key, value):
        if key in self:
            self.move_to_end(key)
        super().__setitem__(key, value)
        if len(self) > self.maxsize:
            oldest = next(iter(self))
            del self[oldest]
```

I'd like to start with the `__init__()` method first.

I know I can type hint `*args` as suggested in [PEP 646: `*args` as a Type Variable Tuple](https://peps.python.org/pep-0646/#args-as-a-type-variable-tuple), but without a kind of variadic generic support for dictionaries, I'm not sure how to type `**kwds`. My thoughts are:

1. `Any`: I'd prefer not to use this as it essentially disables type checking

From the [cpython source](https://github.com/python/cpython/blob/8a0d9a6bb77a72cd8b9ece01b7c1163fff28029a/Lib/collections/__init__.py#L78), I know `OrderedDict` uses `**kwds` in `__init__()`, so I can't manually specify each of the keyword arguments by name.

The `LRU` cache is also a kind of specialized container, so perhaps something like a `Protocol` and `TypeVar` can be used, but I don't know how to make that work for `**kwds`.

Any suggestions?

**Comment by JelleZijlstra** [comment, marked as answer]
I don't think PEP 646 is useful here. The `*args`/`**kwargs` are forwarded to the OrderedDict constructor, which appears to be the same as the dict constructor (according to typeshed). So a precise type signature would simply duplicate the `dict.__init__` signature: https://github.com/python/typeshed/blob/8a47a28e768c577385d6232d4afea496b60403e7/stdlib/builtins.pyi#L1022.

**Reply by adam-grant-hendry** [reply]
I didn't see the `dict.__init__` signature! Thank you! I will study the builtins.

**Reply by adam-grant-hendry** [reply]
@JelleZijlstra So is the best way to type hint the `__init__` this:

```python
from __future__ import annotations

from collections import OrderedDict
from typing import Iterable, TypeVar, overload

from _typeshed import SupportsKeysAndGetItem

_KT = TypeVar('_KT')
_VT = TypeVar('_VT')

class LRU(OrderedDict):

    @overload
    def __init__(self) -> None:
        ...

    @overload
    def __init__(self: dict[str, _VT], **kwargs: _VT) -> None:
        ...

    @overload
    def __init__(self, __map: SupportsKeysAndGetItem[_KT, _VT]) -> None:
        ...

    @overload
    def __init__(
        self: dict[str, _VT], __map: SupportsKeysAndGetItem[str, _VT], **kwargs: _VT
    ) -> None:
        ...

    @overload
    def __init__(self, __iterable: Iterable[tuple[_KT, _VT]]) -> None:
        ...

    @overload
    def __init__(
        self: dict[str, _VT], __iterable: Iterable[tuple[str, _VT]], **kwargs: _VT
    ) -> None:
        ...

    @overload
    def __init__(self: dict[str, str], __iterable: Iterable[list[str]]) -> None:
        ...

    def __init__(
        self: dict[str, _VT] | dict[str, str],
        __map_or_iterable: SupportsKeysAndGetItem[_KT, _VT]
        | SupportsKeysAndGetItem[str, _VT]
        | Iterable[tuple[_KT, _VT]]
        | Iterable[tuple[str, _VT]]
        | Iterable[list[str]]
        | None = None,
        **kwargs: _VT | None,
    ) -> None:
        # Put implementation here...
        pass
```

Oy vey that's a lot of code...
